### PR TITLE
fix: update typescript typings definition file

### PR DESCRIPTION
**Pull request by acheronfail**
**What issue does this pull request resolve?**

There are some issues with the TypeScript typings, especially around the exported classes. Users have had issues with it as well as described in this comment: https://github.com/epoberezkin/ajv/issues/568#issuecomment-405754868

Basically, this issue is that if you try and check if an `error` is an instance of `Ajv.ValidationError` it won't work.

Some steps to reproduce:

```ts
try {
  const validate = ajv.compile(schema); // schema with $async
  await validate(data);
} catch (err) {
  if (!(err instanceof Ajv.ValidationError)) { // <-- TS ERROR HERE
    throw err;
  }
}
```

You get this TypeScript error:

```txt
[ts] The right-hand side of an 'instanceof' expression must be of type 'any' or of a type assignable to the 'Function' interface type.
```

This is because the types don't correctly export the error classes.

**What changes did you make?**

This PR updates the TS typings definitions so that this error is resolved.
When making this change I followed the guides found at: https://github.com/typings/typings/blob/master/docs/examples.md#function--utility which are provided by the TypeScript authors.

**Basically, the changes are:**

* export `declare function ajv(options?: ajv.Options): ajv.Ajv;`
* add `export` to all interfaces and types
* move the error classes into the main namespace and export them

### CODE REVIEW

File: lib/ajv.d.ts [hunk 1]
```diff
@@ -1,13 +1,9 @@
-declare var ajv: {
-  (options?: ajv.Options): ajv.Ajv;
-  new (options?: ajv.Options): ajv.Ajv;
-  ValidationError: ValidationError;
-  MissingRefError: MissingRefError;
-  $dataMetaSchema: object;
-}
+declare function ajv(options?: ajv.Options): ajv.Ajv;
```
Comment (epoberezkin) [inline lib/ajv.d.ts:1]: I think that the usage as constructor may have been lost? Also the issue is to export only ErrorParameters interfaces to use them as type guards - why all interfaces should be exported?

**Review by epoberezkin — CHANGES_REQUESTED** [review]

File: lib/ajv.d.ts (hunk 1 above)
Comment (acheronfail) [reply lib/ajv.d.ts:1]: > I think that the usage as constructor may have been lost?

The `declare function ajv(options?: ajv.Options): ajv.Ajv;` and `export = ajv` in the file ensure that the constructor type isn't lost.

> Also the issue is to export only ErrorParameters interfaces to use them as type guards - why all interfaces should be exported?

Unfortunately, the only way I found to combine the `declare function ajv(...)` type with the `namespace ajv` type was to export all the types inside. Otherwise the types themselves aren't exported as well. 

I tried various ways of fixing the issue, but the only way that worked and didn't break anything was following the typings guide https://github.com/typings/typings/blob/master/docs/examples.md#function--utility here, which says to declare `function` and `namespace` types (where all the types you need from the namespace are exported).

**Comment by epoberezkin** [comment]
replaced by #903
